man/searchcon.1
.TH searchcon 1
.SH NAME
searchcon \- SELinux file context index searching tool
.SH SYNOPSIS
.B searchcon
INDEX [OPTIONS] [EXPRESSION]
.SH DESCRIPTION
.PP
.B searchcon
allows the user to search a file context index created by
.B indexcon
or
.B
apol.
.SH EXPRESSIONS
.P
The following options allow the user to specify which files to print.
A file must meet all specified criteria.
If no expression is provided, all files in the index are printed.
.IP "-t TYPE, --type=TYPE"
Search for files with a context containing the type TYPE.
.IP "-u USER, --user=USER"
Search for files with a context containing the user USER.
.IP "-m RANGE, --mls-range=RANGE"
Search for files with a context with the MLS range of RANGE.
.IP "-p PATH, --path=PATH"
Search for files which include PATH.
.IP "-c CLASS, --class=CLASS"
Search only files of object class CLASS.
.SH OPTIONS
The following additional options are available.
.IP "-R, --regex"
Search using regular expressions instead of exact string matching.
This option does not affect the --object flag.
.IP "-l, --list"
Print a list of types in the index and exit.
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH INFORMATION
If the index file does not contain any MLS ranges and -m was given,
then the search will return nothing.
.PP
Valid object classes are the following:
.PP
file,
dir,
lnk_file,
chr_file,
blk_file,
sock_file,
fifo_file
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2003-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
findcon(1), replcon(1), indexcon(1), apol(1)


NEWS
SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Provides ability to diff levels, categories, and range_transition
    statements.  Provides ability to diff modifications to a user's
    default level and permitted MLS range, assuming the policies are MLS.
    
apol, seaudit, sediffx:

  * Introduces 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  That
    policy list then may be imported into the same tool or any other
    graphical SETools application.  This file also may be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Provides full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Shows MLS diffs.

sediffx:

  * Shows MLS diffs.

  * Provides option to show line numbers that contributed just to a
    specific AV rule's permission by clicking that AV rule's permissions.
    This is in addition to existing sediffx behavior that showed all lines
    associated with a particular rule difference.

  * Provides support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.  Graphical tools have
    a new open policy dialog to select a base module and any number of
    additional modules.

  * Release of RPM packages that are compatible with Fedora Core 5 and
    6.  The spec and support files are in packages/rpm.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

  * New class qpol_module_t to represent a particular policy module
    prior to it being linked into a base policy (qpol_policy_t).

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    fully documented and suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:

  * New diffing algorithm yields significant speed improvements.

  * Can now diff neverallow and role_transition statements.

  * Streamlined results display - should be easier to read.

sechecker:

  * Updated module format and template for ease of extension.


=======================================================
May 1, 2006, Version 2.4

apol:
	File contexts tab now allows for MLS range searching if 
	the loaded database is from a MLS filesystem.
	Policy statistics dialog now shows MLS and ocontexts
	summaries.

libapol:
	Added support for loading base policies containing optionals.
	Added support for searching range transitions containing
	attributes.

libseaudit:
	Bugfix to support parsing FC5-style audit logs.

seaudit:
	Added date filters.

secmds:
	Added support to indexcon and searchcon for MLS filesytems.
	Added support to findcon and replcon for MLS filesystems.
	
sechecker:
	Added incomplete network access (inc_net_access) module.
	Added unreachable domains (unreachable_doms) module.
	Added impossible range transitions (imp_range_trans) module.

sesearch:
	Allow user to search range transitions by attributes and
	indirect matching.
	Added RBAC searching.


=======================================================
January 23, 2006, Version 2.3

apol:
	added new MLS components tab for sensitivities, 
	levels, and categories.
	changed users tab to support ranges and default 
	levels.
	added range transition tab for searching range
	transition rules.

secmds/searchcon.c
/**
 * @file
 * Search the SQLite3 database generated by indexcon or apol for
 * entries that match a given SELinux file context.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 *  Copyright (C) 2003-2007 Tresys Technology, LLC
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *  searchcon: a tool for searching SELinux filesystem databases
 */

#include <config.h>

/* libsefs */
#include <sefs/fsdata.h>
/* standard library includes */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/* command line parsing commands */
#include <getopt.h>

#define COPYRIGHT_INFO "Copyright (C) 2003-2007 Tresys Technology, LLC"

static struct option const longopts[] = {
	{"type", required_argument, NULL, 't'},
	{"user", required_argument, NULL, 'u'},
	{"mls-range", required_argument, NULL, 'm'},
	{"path", required_argument, NULL, 'p'},
	{"list", no_argument, NULL, 'l'},
	{"regex", no_argument, NULL, 'r'},
	{"object", required_argument, NULL, 'o'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'v'},
	{NULL, 0, NULL, 0}
};

static void sefs_double_array_print(char **array, int size)
{
	int i;
	for (i = 0; i < size; i++) {
		printf("%s\n", array[i]);
	}

}

static void sefs_search_keys_ret_print(sefs_search_ret_t * key)
{
	sefs_search_ret_t *curr = NULL;

	/* walk the linked list  */
	curr = key;
	if (curr == NULL) {
		printf("No results\n");
	}
	while (curr) {
		if (curr->context)
			printf("%s\t", curr->context);
		if (curr->object_class)
			printf("%s\t", curr->object_class);
		if (curr->path)
			printf("%s", curr->path);
		printf("\n");
		curr = curr->next;
	}
}

void usage(const char *program_name, int brief)
{
	int size;
	char **array = NULL;
	printf("Usage: %s INDEX [OPTIONS] [EXPRESSION]\n\n", program_name);
	if (brief) {
		printf("\tTry %s --help for more help.\n\n", program_name);
		return;
	}
	printf("Search a file context index for matching files.\n\n");
	printf("  -t TYPE,  --type=TYPE            search for contexts with type TYPE\n");
	printf("  -u USER,  --user=USER            search for contexts with user USER\n");
	printf("  -m RANGE, --mls-range=RANGE      search for contexts with MLS range RANGE\n");
	printf("  -p PATH,  --path=PATH            search for files in PATH\n");
	printf("  -o CLASS, --object=CLASS         search for files of object class CLASS\n");
	printf("\n");
	printf("  -r, --regex                      search using regular expressions\n");
	printf("  -l, --list                       list types in the snapshot\n");
	printf("  -h, --help                       print this help text and exit\n");
	printf("  -v, --version                    print version information and exit\n");
	printf("\n");
	printf("If the index file does not contain any MLS ranges and -m was given,\n");
	printf("then the search will return nothing.\n");
	printf("\n");
	printf("Valid object classes include:\n");
	array = sefs_get_valid_object_classes(&size);
	sefs_double_array_print(array, size);
	sefs_double_array_destroy(array, size);
}

int main(int argc, char **argv, char **envp)
{
	char *filename = NULL;
	int optc = 0, list_sz = 0, list = 0;
	sefs_filesystem_db_t fsdata;
	sefs_search_keys_t search_keys;
	char **list_ret = NULL;
	const char **holder = NULL;

	memset(&search_keys, 0, sizeof(search_keys));

	while ((optc = getopt_long(argc, argv, "t:u:m:p:o:rlhv", longopts, NULL)) != -1) {
		switch (optc) {
		case 't':	       /* type */
			if ((holder =
			     (const char **)realloc(search_keys.type, sizeof(char *) * (search_keys.num_type + 1))) == NULL) {
				printf("Out of memory\n");
				return 1;
			}
			search_keys.type = holder;
			search_keys.type[search_keys.num_type] = optarg;
			search_keys.num_type++;
			break;
		case 'u':	       /* user */
			if ((holder =
			     (const char **)realloc(search_keys.user, sizeof(char *) * (search_keys.num_user + 1))) == NULL) {
				printf("Out of memory\n");
				return 1;
			}
			search_keys.user = holder;
			search_keys.user[search_keys.num_user] = optarg;
			search_keys.num_user++;
			break;
		case 'm':	       /* MLS range */
			if ((holder =
			     (const char **)realloc(search_keys.range, sizeof(char *) * (search_keys.num_range + 1))) == NULL) {
				printf("Out of memory\n");
				return 1;
			}
			search_keys.range = holder;
			search_keys.range[search_keys.num_range] = optarg;
			search_keys.num_range++;
			break;
		case 'p':	       /* path */
			if ((holder =
			     (const char **)realloc(search_keys.path, sizeof(char *) * (search_keys.num_path + 1))) == NULL) {
				printf("Out of memory\n");
				return 1;
			}
			search_keys.path = holder;
			search_keys.path[search_keys.num_path] = optarg;
			search_keys.num_path++;
			break;
		case 'o':	       /* object */
			if ((holder =
			     (const char **)realloc(search_keys.object_class,
						    sizeof(char *) * (search_keys.num_object_class + 1))) == NULL) {
				printf("Out of memory");
				return 1;
			}
			search_keys.object_class = holder;
			search_keys.object_class[search_keys.num_object_class] = optarg;
			search_keys.num_object_class++;
			break;
		case 'l':	       /* list */
			list = 1;
			break;
		case 'r':	       /* regex */
			search_keys.do_type_regEx = 1;
			search_keys.do_user_regEx = 1;
			search_keys.do_range_regEx = 1;
			search_keys.do_path_regEx = 1;
			break;
		case 'h':	       /* help */
			usage(argv[0], 0);
			exit(0);
		case 'v':	       /* version */
			printf("searchcon %s\n%s\n", VERSION, COPYRIGHT_INFO);
			exit(0);
		default:
			usage(argv[0], 1);
			exit(1);
		}
	}
	if (argc - optind > 1 || argc - optind < 1) {
		usage(argv[0], 1);
		exit(-1);
	} else
		filename = argv[optind];

	if (sefs_filesystem_db_load(&fsdata, filename) == -1) {
		fprintf(stderr, "sefs_filesystem_data_load failed\n");
		return -1;
	}

	if (list == 1) {
		if ((list_ret = sefs_filesystem_db_get_known(&fsdata, SEFS_TYPES, &list_sz)) != NULL) {
			sefs_double_array_print(list_ret, list_sz);
			sefs_double_array_destroy(list_ret, list_sz);
		}
	} else {
		sefs_filesystem_db_search(&fsdata, &search_keys);
		sefs_search_keys_ret_print(search_keys.search_ret);
		sefs_search_keys_ret_destroy(search_keys.search_ret);
	}

	free(search_keys.user);
	free(search_keys.type);
	free(search_keys.path);
	free(search_keys.object_class);

	return 0;
}
